### ISSUE TITLE
How to debug an application generated by cement?

### ISSUE THREAD

**Issue by xray1111** [datafolklabs/cement#547]
# Cement Issue Reporting and Feature Requests

Please include any of the following in bug/issue reports where it is relevant:

**System Information**

- Cement Version: 3.0.0
- Python Version: 3.6.6
- Operating System and Version: Mac Os 10.13
- IDE: vscode/pycharm

I generated a application using cement, and I noticed that the entry point of the application is defined by a auto-generated script located at env/bin. the script can run properly but can't be stepped into, either the main.py. I just wonder how to start the debugging work. 

Thank you very much!

**Comment by xray1111** [comment]
Sorry for my poor english.

I mean when open a source project with python IDE such as Pycharm or vscode,  normally I could set some break point and step into or over the code. But I don’t know why I couldn’t debug the code generated by cement. If I run main.py directly, it report an error means couldn’t import library from current directory, and if I debug the app script generated in env/bin, also it report another error. Now I’m using my phone to input these information, I can not paste the exact error. I will copy the error when I get to my computer. Thanks very much!

**Comment by xray1111** [comment]
@derks maybe I could use these two picture to explain what I mean: the first one is the error when I was trying to debug the main.py file, the second pic is the error when I try to  debug the auto-generated application startup script.
<img width="1421" alt="2018-12-18 10 23 32" src="https://user-images.githubusercontent.com/13399784/50160509-852c1a00-0314-11e9-93e6-86a6e5aa1fed.png">
<img width="1429" alt="2018-12-18 10 25 15" src="https://user-images.githubusercontent.com/13399784/50160555-a260e880-0314-11e9-8814-48a257154935.png">

**Comment by ethet** [comment]
I don't have a solution for you, though I wanted to offer my commiseration.

It was my original ambition to simply profile my program to debug a responsiveness/slowness issue (with cProfile or vprof) and have spent the last two days lost in a sea of misery surrounding Python `import` confusion. I have read so many articles and tried so many things, yet I still can't find a way 'into' Cement from the 'outside'.

I feel like I have been faced with this set of problems before (it always feels the worst when you within it), so I am blaming myself and my own ignorance, though I did want to let you know that you aren't alone :)

**Comment by derks** [comment]
@xray1111 @ethet thank you for the feedback and info. I will plan on digging into this after the new year and try to understand it better and figure out a possible solution.

**Comment by ethet** [comment]
Say you build a completely fresh Cement app using [the beginner tutorial](https://docs.builtoncement.com/getting-started/beginner-tutorial). Using that as the base, how would you go about profiling it with [cProfile](https://docs.python.org/3.7/library/profile.html#module-cProfilel) or [vprof](https://github.com/nvdv/vprof)? There just doesn't seem to be a way 'in'. 

If you find that you can easily profile the app, let us know how! The method you use will probably be extraordinarily helpful, as this would also be applicable to PDB, i.e. a way to run and debug the app from `__main__` with all imports resolving.

Though let me put this out there: I believe that true 'problem' is that `import` in Python is an absolute, total nightmare. Your framework is so unbelievably fun to work with, I absolutely love it and it has provided me with an unparalleled sense of accomplishment—in large part because of handlers, controllers, and plugins— though I have to think that part of this wonder results in something which dumbfounds debuggers because of how insane Python imports are.

I want to be shown that I am idiot (because I am), so if you can profile a fresh Cement app from your end—with your intimate knowledge of the framework—show us how and we can mark this thing as 'Solved' :)

**Comment by derks** [comment]
@ethet I'm happy to figure this out.  On your last comment, "There just doesn't seem to be a way 'in'"... was just wondering if you could provide an example of what you've tried (that doesn't work).  That would give me a starting point, rather than going off the rails and down a completely wrong path that isn't helpful here.  FWIW, I've never done any profiling with cProfile or vprof... so, please know that nobody here will be showing you that you're an idiot by any means. ;)  There are some nuances to the framework... and I'm sure room for improvements to make profiling easier (or possible?).

@xray1111 has some examples in the images provided... if you could provide a working example using a Cement tutorial project (ex: `cement generate project -D myapp`), that would be helpful.

**Comment by ethet** [comment]
Let's use cProfile because it is in the standard library...

Assume I created an app with the beginner tutorial and accepted all of the defaults (MyApp) and changed into that directory... 

    python -m cProfile -o prof.dat -m myapp
    ImportError: No module named myapp.__main__; 'myapp' is a package and cannot be directly executed

This is always an issue (we can fix that later :), so we'll create \.__main\__.py and load main():

     cd myapp && touch __main__.py

In __main__.py we run `main()`:

    from main import main
    main()

And here is where everything explodes, no matter how you went about creating your Cement app, by whatever method or means, this erupts:

    File "~/myapp/main.py", line 4, in <module> from .core.exc import MyAppError
    ImportError: attempted relative import with no known parent package

Attempting to solve this by editing Cement at a foundational level only leads to more misery... That is why I believe your intimate knowledge of the framework is the only way to remedy this issue.

**Comment by derks** [comment]
@ethet this is a big help, thank you.  I'll test this out and see what I come up with.

At first glance, there's two high level issues:

- cProfile doesn't like working with setuptools console scripts?
- cProfile doesn't like relative imports called from inside the `__main__` script... 


My first attempt will be to remove the relative imports in `myapp`...  I assume (hope) this is only relevant becuase of the way cProfile loads the `__main__`... becuase cement is 100% relative imports as well, and that would be more problematic to change.

**Comment by derks** [comment]
My first swipe... this worked, and generated a `prof.dat` (not sure what to do with it yet):

```
$ python -m cProfile -o prof.dat -m myapp.main
```

**Comment by derks** [comment]
@ethet this is all completely valid feedback for me, and I honestly appreciate someone "wasting my time" more than anyone being frustrated with the framework and not saying something about it.

I'm keeping this open as a note to add a doc on debugging so that it can help anyone else along in the future.

**Comment by derks** [comment]
@ethet to be honest, after profiling your app(s) I'd really love to know if there are any red flags in the framework runtime that you could help identify and improve.

**Comment by ethet** [comment]
It's funny... This whole thing came about because I love my Cement app so much that I created a REST API for it so that I could sync it between my computers, and then I 'needed' it on my phone so I began creating a web app to interface with said API (so that I can have my Cement app everywhere). 

Though when I use the web app the responsiveness is instant, immediate, so fast that I can't detect the internet between us (a Flask API serving Postgres data to vanilla JavaScript in the browser). Though on my local machine, with the Cement app loading right beneath me on a 4Ghz CPU with tons of memory, it takes about a second for the program to respond (even if I turn off syncing). 

I am doing all kinds of nutzo stuff with Colorama and Jinja filters, though I can't help but assume that I doing something wrong with setup hooks and the way in which I am using (abusing?) handlers and plugins. Though it is all of these wonderful affordances which make Cement so amazingly fun to work with... so I need the profiling to show me where I have introduced bottlenecks, because apparently I have.

Thanks again for all of your help... You have no idea how much I love what you have created and how high I hold you in regard. I have been through the source code line by line when debugging my own faults (and trying to find a way to hijack Cement's Jinja implementation :) and have learned an amazing amount from you by doing so.

Thank you for putting this out there for us to use... There is something exhilarating about Cement, especially when one discovers all of the 'hidden' features (like the honesty of a one-to-one compatibility with argparse with things like {'nargs': '+'}—that one turned my head around when it worked). The framework feels carefree yet restrained at the extremes, and I love what you have done... especially for something so prosaic and archaic as command line apps. Who could have imagined that applying modern methodologies to the CLI could feel so punk rock?

**Comment by derks** [comment]
@ethet I appreciate all of the kind words.  It has been nearly 10 years now, with 3 complete redesigns... I would shudder to have you go through Cement 0.X line by line...  haha  But I've always been passionate about CLI and backend development which is obviously where Cement was born out of.

Regarding running Flask vs. direct on command line... Is that a separate Flask app shelling out to Cement, or is that Flask integrated *into* Cement (which I've done).  The reason I ask is, if Flask is integrated into Cement then you are taking all of the setup cost only once and then everything is running from memory (warm hits).  When you call Cement via command-line it takes the full setup hit every time (cold hits).

I've had issues with startup time as well, and have always had plans to profile and improve it... but never found the time and most of the utility I write on Cement are automation/non-interactive so I don't feel it in production.

I've had thoughts of crazy/interesting ways of speeding up the cold start... like, pickling/caching parts of the app for a 'warm' startup if the config hasn't changed at all... but not sure that is smart or even a normal thing to think of at all.  I think time would be better spent getting the interactive shell feature going so that the app is loaded once and stays loaded in it's own shell. It's doable, and [there are issues open for it](https://github.com/datafolklabs/cement/issues/276), but I just haven't had the time to decide how I want to do that.

---

FWIW, I had the idea a while ago about building an extension that would automatically integrate and expose an entire Cement app via Flask so that the only thing you would need to do is include the extension, and run `myapp serve` or something like that.  Maybe open a new issue if you'd like to see something like that... or have any insight into how you did it with your use-case.

**Comment by ethet** [comment]
No, the Flask API is completely separate. I have created a Cement plugin which syncs native data with a remote Flask server upon the successful execution of subcommands in Cement controllers (the ease of this activity with plugins/handlers speaks volumes, and kinda exemplifies why I love Cement so much). The web app then connects to that same remote server, without any interaction with Cement per se, though mimicking a Cement CLI with [jQuery Terminal](https://terminal.jcubic.pl)... Yes, I love Cement so much I am actually *EMULATING IT* in JavaScript...

But the JavaScript version is like a million times faster than my native Cement app... And that is were everything began going off the rails and how I ended up here.

I have been working on this the past couple of days and am now wondering if this is a Python startup problem. I haven't yet dug into the profiling... That is going to be a labor of love that I plan on beginning in the morning. But I have a relatively comprehensive pytest suite and I noticed something: It takes a slight moment to start and then rips through the tests.

I have a Qt for Python app that I made (with remote syncing) and it is similar... It takes a moment to load but once everything is in memory it is super responsive. 

Though Cement has to completely rebuild itself upon every command... or so it seems, and thus seems to forever be a victim of Python's slow start.

Though this is complete conjecture... I haven't even done the work yet to establish where my problem lies.

Thank you for the conversation, it is kinda thrilling :)

**Comment by xray1111** [comment]
@derks @ethet Sorry for late response!
Actually I found that the auto generated entry code under env/bin could step into and debug from there using Pycharm, but still failed with vscode, maybe I config some thing incorrectly. If @derks could write some document on how to debug under cement framework, I think it would be very helpful to the beginners.
Thanks again for your help!

**Comment by xray1111** [comment]
Hi, @derks , I was recently considering how to use cement framework to construct a flask rest api application. Since you mentioned flask earlier in this topic, I append this question in the thread.

I really want my program be uniformed under some kind of specific framework, such as cement, but seems like it's difficult to combine these 2 things. Any idea?

**Comment by derks** [comment]
@xray1111 do you mean simply having a Cement CLI app for commands/arguments... and then running a Flask app from a sub-command... or do you mean mapping Cement CLI sub-commands to Flask Routes some way (or vice-versa)?

**Comment by xray1111** [comment]
I guess the first choice would do? Actually I'm a bit confused cause I'm also a newbie on flask, sorry...

Is there a way to wrap around flask api so that I can use the flask decorator to expose my rest api, and the program can still be structured with cement's style? 

Or, let's say if you are writing some rest api application base on flask, how would you do if you don't want to give up cement framework? Thanks!

**Comment by derks** [comment]
Many years later... I've added a doc on profiling:

- https://docs.builtoncement.com/additional-topics/profiling-with-cprofile


@xray1111 It's been ages... sorry I never replied.

I think there are two things:

1. Using Cement as the base app to provide a command line interface for running a Flask App. For example, using Cement to support configuration files and command line arguments.. but ultimately running Flask and doing everything else the Flask way. I see this as useful for example, `myapp webserver --listen 0.0.0.0 --port 8888` .... essentially you'd be using Cement to capture listen/port/other-config-settings and using thos to pass to `flask.App`.
2. Mapping Cement commands/arguments directly to Flask endpoints. This would be a specialty case and would be a topic all on it's own so I'm going to continue to "punt" on that for now.  Cement was built for CLI ... I'm sure someone could make a "Flask Extension" that would do this, but it's not something I've had motivation to try.
